### ISSUE TITLE
Conflicts in personal preferences

### ISSUE THREAD

**Issue by christianbundy** [fraction/oasis#224]
Was going to post this in a comment in #223 but I think the scope of this discussion is wider than that code. cc: @ciex @cinnamon-bun @jedahan 

---

**What's the problem you want solved?** I'm not sure how to resolve conflicts in personal preferences, like the horizontal layout vs vertical layout, without feeling like I'm trampling all over someone's freshly planted garden. I'm *also* new to this C4 style and I'm concerned that I'm hurting feelings or being hostile/unwelcoming.

@ciex I'm really excited that you've picked this up and I want to support your experiments!

I **love love love** the summaries, but either my brain or my screen size doesn't mesh well with the horizontal layout and it makes Oasis feel less calm for me. I really want @ciex to be able to use it though, so maybe there's a way we can enable that? Maybe we need a way to enable custom CSS? I'm just kind of afraid that any styles that aren't "mainline Oasis" are going to get broken with how fast development is moving along.

**Is there a solution you'd like to recommend?**

Finding a way to steward Oasis in a way that works for everyone. 🧑‍🤝‍🧑🧑‍🤝‍🧑 I guess I just want to check in and see how #219 and #223 feel for everyone because this is new for all of us and I'm feeling self-conscious about opening a PR to revert the horizontal layout. :heart: Thanks for putting up with me.

**Comment by jedahan** [comment]
From what I understand the goal of C4 is to maximize the number of contributors.

I have seen a few different descriptions for the goal of Oasis, including ones that seem to be in conflict with what gets merged. Historically, I have kept quiet about the changes because # of contributors matter more than what I personally would prefer, and would love to continue on that trend. That is to say I will tend to support merging things I personally dislike/think are not improvements if it means we have a healthier community for it. Perhaps one day it'll reach a point where I feel the need to fork, but oasis is so, so early, and I think many forks/options are an okay thing to have before calcification occurs.

Ask again tomorrow, and I might have completely different thoughts :)

**Comment by jedahan** [comment]
One thing of note - with CSS, I think there is room for a single codebase with multiple views. That is to say, for #219 I think there is room for something that is calmer at one size and deeper at another. This could be yet another view like in #223, or perhaps a moon/sun toggle which hides many interface elements.

**Comment by georgeowell** [comment]
I think it's fine. It's just a question of iteration. I would foresee that all the ideas will get mashed together into a beautiful mixture of everyone's efforts. Because the front end of Oasis is so to modify, then people are free to fork it and make their own lunarpunk variations. This will only become easier when the codebase stabilises. 

On those commits specifically, I personally think we have committed to a vertical layout and should proceed with that in mind.

**Comment by georgeowell** [comment]
This is just distant conjecture but perhaps Oasis could have some sort of "App" plug-in functionality like Soapdog was planning with Patchfox where people can add small views and apps that they can toggle on and off.

I'm kind of against this in some ways but it's something to think about and would address that problem.

**Comment by cinnamon-bun** [comment]
I bet this will be a common pattern with C4 - someone contributes a patch, and then it gets adjusted and iterated by several more people before it settles down.

We could suggest that new features go into some kind of scratch space for a while until they get polished.  That could mean making new pages, or using feature flags.

Big-picture though, I'm not sure C4 works very well for user-facing software (as opposed to libraries and modules).  As this article points out: [Why designing for open source can be so difficult](https://gilli.is/journal/why-designing-for-open-source-can-be-so-difficult)

> I think a part of the reason why more designers don't contribute to open source is that often it requires you to rethink the entire product from the ground up and make changes that require someone else to do a lot of work. And maybe even remove features that blocking a good experience. That's actually how I start every project I work on, I dig in and question everything and make sure to strip it down to its bare bones and build it back up with a clear focus.

> I think removing features is possibly one of the hardest things to do for an open source project because there will always be that one guy who really, really, really wants that feature and will be very vocal about it. For a project with a financial incentive, this is an easy decision to make. But for an open source project, this can lead to an endless discussion spanning many months. On top of that, someone may have put a lot of work into building that feature and might feel insulted if it is deleted. All this can lead to a lot of resentment and drain the energy out of the project.

TL;DR - Design can be about simplicity, consistency, and curation of features, but it feels bad to delete other peoples' code.

Solutions?
* Write up design guidelines like "views are a single column flowing top to bottom, indentation represents nesting"
* Add a guideline to C4 rules: new features that change the user experience significantly should go on a new page or behind a feature flag, and might eventually be removed

I used to live in a group house with lots of murals.  Anyone could paint a mural without asking, but also anyone could **remove** a mural without asking.  On the other hand, murals that the house reached consensus about became "protected" and could not be removed without another house meeting.

Maybe new UX experiments eventually graduate from the sandbox through a process like that?

¯\_(ツ)_/¯

**Comment by jedahan** [comment]
> views are a single column flowing top to bottom, indentation represents nesting

While these are currently technically neutral observations, encoding them **as a guideline** discourages exploration, discourse, and ultimately improvement.

---

> ...UX experiments eventually graduate from the sandbox...

Would love to see this happen, and personally think that letting the default branch be experimental is the healthiest thing. I was imagining `moon mode` as the most stripped down core ux (calmer, and protected status), and `sun mode` as illuminating every crack and corner and route someone came along to try.

We can also celebrate the removal of features (colloquially on another project, we celebrate salmon as the default background color for removed code).

---

I'd hope that design guidelines are not prescriptive but descriptive, and specifically around design considerations and prioritization. Here are examples that do not represent my opinion:

Example prioritizations - things we have observed even just a little bit

- We have added views, which gives options for long-term users at the cost of first-time experience
- We have kept routes in a single file, which makes it easy to remove features in the future

Example considerations - things that may aid you when contributing

- How does this contribution change the difficulty of further contributions?
- How does this contribution change the initial expectations when opening oasis?
- How does this contribution change the accessibility of navigating oasis on a laptop?
- How does this contribution change the accessibility of navigating oasis on a phone?
- How does this contribution encourage some kinds of discourse?
- How does this contribution discourage some kinds of discourse?
- How easy is it to remove this contribution?

**Comment by christianbundy** [comment]
Two bits from `contributing.md` that I think are relevant:

> All Public Contracts (APIs or protocols) MUST be documented.

> A patch that modifies a stable Public Contract SHOULD not break existing applications unless there is overriding consensus on the value of doing this.

I think documentation constitutes a public contract, so as long as we're documenting our design language somewhere then I think it makes sense to protect that public contract against unexpected changes. For example, the readme says "No browser JavaScript!", and I wouldn't want to merge any patches that changed that policy without consensus from active maintainers.

Any time we engage in design there are going to be implicit expectations, regardless of an explicit public contract, and if we bump into pain points where those implicit expectations vary between users/developers then I think it would be wise to make those decisions explicit and document them in a public contract.

> moon/sun toggle

Just to chime in on this: I'd really like for Oasis to be cohesive, so I'd be hesitant to add any customization that's anything deeper than superficial (themes, language, etc), but if there's a problem we need to take care of I'd love to look at the solutions we have available.

**Comment by ciex** [comment]
Noncommercial open source software has a tendency to cater a lot to personal preferences, in the sense that often more options and switches are available. Sometimes this can be really cool and liberating, but often it also is hard to maintain and overwhelming, especially for new users (and new contributors).

I like the idea of welcoming change but also creating space for it to move around before settling down. There is a real possibility of new development becoming permanent in open source development not because of consensus but just because nobody else wants to work on that right now. How can we make sure that changes like this are experimented with and moved around by contributors? What could a process be to revert a change like this if it doesn't prove itself?

**Comment by georgeowell** [comment]
Closing as discussion inactive.
